kotlinx-coroutines-io/kotlinx-coroutines-io-jvm/src/main/kotlin/kotlinx/coroutines/experimental/io/internal/CancellableReusableContinuation.kt

```diff
@@ -103,32 +88,33 @@ internal class MutableDelegateContinuation<T : Any> : Continuation<T>, Dispatche
         c.resumeWith(Result.failure(exception))
     }
 
-    private inner class JobRelation(val job: Job) : CompletionHandler, DisposableHandle {
-        private var handler: DisposableHandle = NonDisposableHandle
+    private inner class JobRelation(val job: Job) : CompletionHandler {
+        private var handler: DisposableHandle? = null
 
         init {
+            @UseExperimental(InternalCoroutinesApi::class)
             val h = job.invokeOnCompletion(onCancelling = true, handler = this)
+
             if (job.isActive) {
                 handler = h
             }
         }
 
         override fun invoke(cause: Throwable?) {
-            this@MutableDelegateContinuation.handler.compareAndSet(this, null)
-            dispose()
+            this@CancellableReusableContinuation.jobCancellationHandler.compareAndSet(this, null)
```

Comment:
shouldn't it be protected with if?
```
if (compareAndSet) {
  ... invoke handler and resume
}
```

Reply:
No, it shouldn't. Will rewrite to clarify it